pkgs/build-support/writers/default.nix

```diff
@@ -22,7 +22,30 @@ rec {
       inherit interpreter;
       contentPath = content;
     }) ''
-      echo "#! $interpreter" > $out
+      # On darwin a script cannot be used as an interpreter in a shebang but
+      # there doesn't seem to be a limit to the size of shebang and multiple
+      # arguments to the interpreter are allowed.
+      if [[ -n "${toString pkgs.stdenvNoCC.isDarwin}" ]] && isScript $interpreter
+      then
+        wrapperInterpreterLine=$(head -1 "$interpreter" | tail -c+3)
+        # Get first word from the line (note: xargs echo remove leading spaces)
+        wrapperInterpreter=$(echo "$wrapperInterpreterLine" | xargs echo | cut -d " " -f1)
```

Comment:
I don’t quite understand how this is intended to work.

If `$interpreter` is a binary, it will go through the binary until it finds the first `\n`, right? But binaries can also have \0 at arbitrary positions and such …

The code seems a little confused to me, I can’t see how it would work for all interpreters.

Reply:
At this point the interpreter should never be a binary due to “ isScript $interpreter” on line 28

Reply:
btw. I borrowed the code from `patch-shebangs.sh`

https://github.com/NixOS/nixpkgs/blob/e50c67ad7eefa8e77436fbd0366b69638b1c8713/pkgs/build-support/setup-hooks/patch-shebangs.sh#L51-L53

Reply:
Maybe it should say so in a comment.